pkg/controller/job/job_controller.go

```diff
@@ -1724,7 +1735,16 @@ func ensureJobConditionStatus(list []batch.JobCondition, cType batch.JobConditio
 	return list, false
 }
 
-func isPodFailed(p *v1.Pod, wFinalizers bool) bool {
+func isPodFailed(p *v1.Pod, job *batch.Job, wFinalizers bool) bool {
+	if feature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) && feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) && job.Spec.PodFailurePolicy != nil {
+		// When PodDisruptionConditions is enabled, orphan Pods are marked as Failed.
+		// We can wait for Pods to fully terminate to check the pod failure policy.
+		// TODO(#113855): Stop limiting this behavior to Jobs with podFailurePolicy.
+		// For now, we do so to avoid affecting all running Jobs without the
+		// avaibility to opt-out into the old behavior.
+		return p.Status.Phase == v1.PodFailed ||
```

Comment:
I think it would be good to add an e2e test to show that the PR fixes the issue. I think it could be a similar test to this one: https://github.com/kubernetes/kubernetes/blob/8e48df135318026d5f8a972a96a2ff665889568a/test/e2e/apps/job.go#L188, but instead of looking for DisruptionTarget condition it would look for exit code 137. On master the test would fail as the pod is matched against the policy before the pod is terminated (and thus the exit code 137 is set). Such an e2e test would also prevent a similar regression in the future.